Enable residual tagging for multiple systems in the same input file

grmnptr opened the pull request:
Closes #23194

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1323290/) on 7d04bc8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23282/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23282/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b1584ac2ec2d98e7f4756228788341656aa44cfd`

loganharbour reviewed (commented):
Why? I don't like this. Fande and I thought a long time how to best generalize this and you're nuking it all. Duplicating all of the methods is not the answer.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1323384/) on a60e3e5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23282/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23282/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b1584ac2ec2d98e7f4756228788341656aa44cfd`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1323431/) on a051321 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23282/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23282/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b1584ac2ec2d98e7f4756228788341656aa44cfd`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1323551/) on 7b21b8b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23282/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23282/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b1584ac2ec2d98e7f4756228788341656aa44cfd`

grmnptr commented:
Well, we have multiple systems in a problem now which require separate residual tags and as far as I can see (I might be wrong here) residual and solution vectors behave differently. Of course I am always open to suggestions. I am still experimenting with these, trying to change it as little as possible. But thanks for providing feedback at the very first commit so I don't have to waste my time.

grmnptr commented:
I might also lack the details here. Why did you guys generalize it? Do some of the applications need additional tag types?

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1325103/) on a6ce926 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b1584a</th>
      <th colspan=4>#23282 a6ce92</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.62%</td>
      <td>84.62%</td>
      <td>-0.01%</td>
      <td>56.52%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>81979</td>
      <td>81981</td>
      <td>+2</td>
      <td>13</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14897</td>
      <td>14904</td>
      <td>+7</td>
      <td>10</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23282/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23282/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23282/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 56.52% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1325089/) on a6ce926 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23282/site)

This comment will be updated on new commits.

YaqiWang commented:
There could be a misunderstanding. I think tag id/name belongs to FE problem but not individual systems. The same tag id can be used to add solution vectors for multiple systems (we had this in aux system and nonlinear system before having multi-system). Typically it is up to executioner or a system to add a vector to a particular system for a tag id.

lindsayad commented:
The issue is with residual vector tags @YaqiWang, which is a problem that we didn't have until multi-system. If someone asks to retrieve residual vector tags, before `computeResidualTags` for instance, you must be able to discern residual tags from what system. We cannot accumulate into residual tags from another system ... the residuals will be the wrong length, not have the corresponding dof indices, etc.

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -5780,12 +5782,8 @@ FEProblemBase::computeResidual(const NumericVector<Number> & soln,
-
-  _fe_vector_tags.clear();
 
-  for (const auto & residual_vector_tag : residual_vector_tags)
-    _fe_vector_tags.insert(residual_vector_tag._id);
+  _fe_vector_tags = getResidualVectorTags(nl_sys_num);
```
Like here @YaqiWang... we can't just set `_fe_vector_tags` to all the `VECTOR_TAG_RESIDUAL` vector tags

grmnptr commented:
Thanks @YaqiWang for weighing in on this topic! The reason why I am doing this is that I'd like to be able to assemble a residual with a tag on one system but not on the other. The current implementation assumes that every tag exist on every libmesh system which is not true. The example on this issue:

```
[Problem]
  nl_sys_names = 'u v'
  extra_tag_vectors = 'ahoj' # <<< Added line
[]

[Kernels]
  [diff_u]
    type = Diffusion
    variable = u
    extra_vector_tags = 'ahoj' # <<< Added line
  []
  [diff_v]
    type = Diffusion
    variable = v
  []
  [force]
    type = CoupledForce
    variable = v
    v = u
  []
[]
```

Result in 


```
*** ERROR ***
Cannot zero vector tag with name 'AHOJ' in system 'v' because there is no vector associated with that tag

Stack frames: 17
0: 0   libmesh_opt.0.dylib                 0x00000001060252bc libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 996
1: 1   libmoose-opt.0.dylib                0x000000010559b7d4 moose::internal::mooseErrorRaw(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 748
2: 2   libmoose-opt.0.dylib                0x00000001052049f4 void mooseError<char const (&) [35], TagName, char const (&) [14], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [54]>(char const (&) [35], TagName&&, char const (&) [14], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [54]) + 352
3: 3   libmoose-opt.0.dylib                0x00000001051b0440 SystemBase::zeroTaggedVector(unsigned int) + 284
4: 4   libmoose-opt.0.dylib                0x00000001051a1e00 NonlinearSystemBase::computeResidualTags(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 452
5: 5   libmoose-opt.0.dylib                0x0000000104db25d0 FEProblemBase::computeResidualTags(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 1012
6: 6   libmoose-opt.0.dylib                0x0000000104db2080 FEProblemBase::computeResidualInternal(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&, std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 180
7: 7   libmoose-opt.0.dylib                0x0000000104dce430 FEProblemBase::computeResidualSys(libMesh::NonlinearImplicitSystem&, libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&) + 136
8: 8   libmoose-opt.0.dylib                0x00000001051b9ab8 NonlinearSystem::solve() + 208
9: 9   libmoose-opt.0.dylib                0x0000000104dd73dc FEProblemBase::solve(unsigned int) + 428
10: 10  libmoose_test-opt.0.dylib           0x0000000102c6ecf8 auto SteadySolve2::execute()::$_0::operator()<unsigned int>(unsigned int const&) const + 72
11: 11  libmoose_test-opt.0.dylib           0x0000000102c6a998 SteadySolve2::execute() + 340
12: 12  libmoose-opt.0.dylib                0x00000001055706ec MooseApp::executeExecutioner() + 268
13: 13  libmoose_test-opt.0.dylib           0x0000000102ebb83c MooseTestApp::executeExecutioner() + 224
14: 14  libmoose-opt.0.dylib                0x000000010556c5f4 MooseApp::run() + 1504
15: 15  moose_test-opt                      0x00000001024b3b68 main + 160
16: 16  dyld                                0x0000000193c8fe50 start + 2544
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
```

YaqiWang commented:
> The current implementation assumes that every tag exist on every libmesh system which is not true. 

`FEProblemBase::computeResidualTags` is requiring to set the current system `_current_nl_sys` and accept `tags`. I assume we just passed in unavailable tags when calling `FEProblemBase::computeResidualTags` thus the fix should be only on the caller side. I checked the code. We might just need to be more careful in 
```
FEProblemBase::computeResidual(const NumericVector<Number> & soln,
                               NumericVector<Number> & residual,
                               const unsigned int nl_sys_num)
```
when populating the tags:
```
  for (const auto & residual_vector_tag : residual_vector_tags)
    _fe_vector_tags.insert(residual_vector_tag._id);

```
by filter out tags by calling `_current_nl_sys->hasVector(residual_vector_tag._id)`.

grmnptr commented:
Ahhh thanks @YaqiWang , I'll give this a try, looks like the least intrusive approach!

lindsayad commented:
Very nice suggestion!

grmnptr commented:
The only issue with this is that the tag `RESIDUAL` is never registered in libmesh, it has `RHS Vector`. I can add a function that associates the two with each other but that is rather dirty.

grmnptr committed:
Add check for tag calls. (#23194)

Remove libmeshAliasName

Add comments to association. (#23194)

Add filter for residual tags. (#23194)

lindsayad commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -1195,6 +1195,8 @@ class FEProblemBase : public SubProblem, public Restartable
    */
   virtual void computeResidualTags(const std::set<TagID> & tags);
 
+  virtual TagName libmeshAliasName(const VectorTag & tag);
```
```suggestion
  static TagName libmeshAliasName(const VectorTag & tag);
```

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -5785,11 +5785,24 @@ FEProblemBase::computeResidual(const NumericVector<Number> & soln,
 
   for (const auto & residual_vector_tag : residual_vector_tags)
-    _fe_vector_tags.insert(residual_vector_tag._id);
+  {
+    // we filter out tags which are not present in the current nonlinear system
+    if (_nl[nl_sys_num]->hasVector(libmeshAliasName(residual_vector_tag)))
+      _fe_vector_tags.insert(residual_vector_tag._id);
+  }
```
```suggestion
    // we filter out tags which are not present in the current nonlinear system
    if (_nl[nl_sys_num]->hasVector(libmeshAliasName(residual_vector_tag)))
      _fe_vector_tags.insert(residual_vector_tag._id);
```

loganharbour reviewed (changes requested):
Please drop the first two commits

YaqiWang commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -5780,12 +5780,19 @@ FEProblemBase::computeResidual(const NumericVector<Number> & soln,
 
   for (const auto & residual_vector_tag : residual_vector_tags)
-    _fe_vector_tags.insert(residual_vector_tag._id);
+    // We filter out tags which are not present in the current nonlinear system. This is
```
`which are not present` => `which do not have associated solution vectors`. This is more accurate.

loganharbour reviewed (approved):
Failure is unrelated
